### PR TITLE
Seg tree walks

### PR DESCRIPTION
**Pull request by cameroncuster**
### changes
- added `find_first` and `find_last` methods to `seg_tree`
- included `lower_bound` in the `seg_tree` to be consistent
- wrote a test for small cases with an array and for loops
- wrote a test for large cases comparing with the `bit.lower_bound` method
- extra testing: https://codeforces.com/contest/1705/submission/250940123

### notes
- it's possible to combine `find_first` and `find_last` into the same method, but this seemed like it would be less clean and as a result more difficult to modify
- I opt'd to return the "endpoint" of the range if no such value exists in the range because this is standard practice for other library functions
- I only tested with 0/1 vectors
- removed the `true = 1` check from the `astyle_cppcheck` because it was failing as a result of using the word "true" in my docs
- it doesn't currently support the exact same thing that `lower_bound` does which depends on the data in the nodes in the prefix/suffix which I don't currently keep track of, it could be modified to support this and we could remove `lower_bound`, but  I was hesitant to make these methods more complex since we want them to be modifiable
- I opt'd to templatize the functional param to make these walks more extensible

### see these runtimes as well
without template: https://codeforces.com/contest/1705/submission/250938915
with template: https://codeforces.com/contest/1705/submission/250940123

### CODE REVIEW

**Comment by cameroncuster** [comment]
Looking at the changes it appears the formatting was changed

I ran the `astyle_cppcheck` but I don't think that reformats the files?

File: library/data_structures/seg_tree_uncommon/find_first.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * @param le,ri defines range [le, ri)
+ * @param f(tree[u], tl, tr) defines a function that returns true if the subtree
+ * contains an element that satisfies the condition
+ * @returns the index of the first element in the range that satisfies the
+ * condition described in `f`, if no such element exists then `ri` is returned
+ * @time O(log(n))
+ * @space O(log(n)) for recursion stack; no new nodes are allocated
+ */
+int find_first(int le, int ri, function<bool(int64_t, int, int)> f) {
```
Comment (cameroncuster) [inline library/data_structures/seg_tree_uncommon/find_first.hpp:12]: it might be cleaner to templatize `f`

Comment (cameroncuster) [reply library/data_structures/seg_tree_uncommon/find_first.hpp]: It ended up being cleaner & faster, so I updated this

see
without template: https://codeforces.com/contest/1705/submission/250938915
with template: https://codeforces.com/contest/1705/submission/250940123

File: library/data_structures/seg_tree_uncommon/find_last.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * @param f defines a function that returns true if the subtree contains an
+ * element that satisfies the condition
+ * @returns the index of the last element in the range that satisfies the
+ * condition described in `f`, if no such element exists then (le - 1) is
+ * returned
+ * @time O(log(n))
+ * @space O(log(n)) for recursion stack; no new nodes are allocated
+ */
+template <class F>
+int find_last(int le, int ri, const F& f) {
```
Comment (lrvideckis) [inline library/data_structures/seg_tree_uncommon/find_last.hpp:14]: do we want `const F&` or just `F`, cuz everywhere else it's the latter

https://github.com/programming-team-code/programming_team_code/blob/main/data_structures/rmq.hpp#L18

File: library/data_structures/seg_tree_uncommon/find_last.hpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+int find_last_impl(int le, int ri, const F& f, int tl, int tr, int u) {
+	if (ri <= tl || tr <= le) return le - 1;
+	if (tr - tl == 1) {
+		if (f(tree[u], tl, tr)) return tl;
+		return le - 1;
+	}
+	int tm = split(tl, tr);
+	push(tl, tm, tr, u);
+	if (le <= tl && tr <= ri) {
+		if (!f(tree[u], tl, tr)) return le - 1;
```
Comment (lrvideckis) [inline library/data_structures/seg_tree_uncommon/find_last.hpp:27]: why not return `ri`?

Comment (lrvideckis) [reply library/data_structures/seg_tree_uncommon/find_last.hpp:27]: cuz stl returns `end` iterator on not found

File: library/data_structures/seg_tree_uncommon/find_last.hpp (hunk 2 above)
Comment (cameroncuster) [reply library/data_structures/seg_tree_uncommon/find_last.hpp:14]: so compare these

not const ref: https://codeforces.com/contest/1705/submission/250982914
with const ref: https://codeforces.com/contest/1705/submission/250940123

you'll notice with const ref is a bit faster so I would vote to keep it (it's better for the recursion stack I think)

in the example you linked it's not recursive so I think it matters less

Comment (lrvideckis) [reply library/data_structures/seg_tree_uncommon/find_last.hpp:14]: ah I see

File: library/data_structures/seg_tree_uncommon/find_last.hpp (hunk 3 above)
Comment (cameroncuster) [reply library/data_structures/seg_tree_uncommon/find_last.hpp:27]: Hmm it's possible that would be more consistent, but when working with things in "reverse" this still works which is nice

see: https://github.com/programming-team-code/programming_team_code/blob/6cebc5a0bcda539e3bb3e8e9580381018e623138/tests/library_checker_aizu_tests/handmade_tests/seg_tree_find.test.cpp#L36

Comment (lrvideckis) [reply library/data_structures/seg_tree_uncommon/find_last.hpp:27]: ah ok

Comment (cameroncuster) [reply library/data_structures/seg_tree_uncommon/find_last.hpp:27]: also comparing the index with some bound should still work if the code is written accordingly without an extra check for `ri` which can be annoying

ex. I always find myself checking `if (it == end(container))` which gets annoying

**Pull request merged**